text/0000-crt-static.md

```diff
@@ -0,0 +1,371 @@
+```
+
+Through this method build scripts will be able to learn how the C standard
+library is being linked.  This is crucially important for the MSVC target where
+code needs to be compiled differently depending on how the C library is linked.
+
+This feature ends up having the added benefit of informing build scripts about
+selected CPU features as well. For example once the `target_feature` `#[cfg]`
+is stabilized build scripts will know whether SSE/AVX/etc are enabled features
+for the C code they might be compiling.
+
+After this change, the gcc-rs crate will be modified to check for the
+`CARGO_CFG_TARGET_FEATURE` directive, and parse it into a list of enabled
+features. If the `crt-static` feature is not enabled it will compile C code on
+the MSVC target with `/MD`, indicating dynamic linkage. Otherwise if the value
+is `static` it will compile code with `/MT`, indicating static linkage. Because
+today the MSVC targets use dynamic linkage and gcc-rs compiles C code with `/MD`,
+gcc-rs will remain forward and backwards compatible with existing and future
+Rust MSVC toolchains until such time as the the decision is made to change the
+MSVC toolchain to `+crt-static` by default.
+
+### Lazy link attributes
+
+The final feature that will be added to the compiler is the ability to "lazily"
+interpret the linkage requirements of a native library depending on values of
+`cfg` at compile time of downstream crates, not of the crate with the `#[link]`
+directives. This feature is never intended to be stabilized, and is instead
+targeted at being an unstable implementation detail of the `libc` crate linked
+to `std` (but _not_ the stable `libc` crate deployed to crates.io).
+
+Specifically, the `#[link]` attribute will be extended with a new argument
+that it accepts, `cfg(..)`, such as:
+
+```rust
+#[link(name = "foo", cfg(bar))]
+```
+
+This `cfg` indicates to the compiler that the `#[link]` annotation only applies
+if the `bar` directive is matched. This interpretation is done not during
+compilation of the crate in which the `#[link]` directive appears, but during
+compilation of the crate in which linking is finally performed. The compiler
+will then use this knowledge in two ways:
+
+* When `dllimport` or `dllexport` needs to be applied, it will evaluate the
+  final compilation unit's `#[cfg]` directives and see if upstream `#[link]`
+  directives apply or not.
+
+* When deciding what native libraries should be linked, the compiler will
+  evaluate whether they should be linked or not depending on the final
+  compilation's `#[cfg]` directives and the upstream `#[link]` directives.
+
+### Customizing linkage to the C runtime
+
+With the above features, the following changes will be made to select the
+linkage of the C runtime at compile time for downstream crates.
+
+First, the `libc` crate will be modified to contain blocks along the lines of:
+
+```rust
+cfg_if! {
+    if #[cfg(target_env = "musl")] {
```

Comment:
Yes if support is added to get that working, we can encode that here. The specifics of that, however, are out of scope of this RFC. It's intended though that there's enough plumbing here that one could imagine:

```
RUSTFLAGS='-C target-feature=+crt-static' cargo build --target x86_64-unknown-linux-gnu
```

to one day work!